### PR TITLE
feat(receive): Add tests for InputPair constructors Closes #583

### LINKED ISSUE #583: InputPair's lack of a constructor seems to cause mistakes
@kumulynja mentioned to me that 

> [in] the last step of broadcasting the payjoin transaction I get the following error:
> 
> BdkError.electrum(field0: Electrum server error: "sendrawtransaction RPC error -26: mandatory-script-verify-flag-failed (Signature must be zero for failed CHECK(MULTI)SIG operation)")
> 
> The proposal psbt is finalized when signing with the sender before broadcasting, and I think I am using the same signing options as in the current version, so not sure what to change there.

Turned out @kumulynja was passing empty lists to the `psbt::Input`'s `redeemScript` and `witnessScript` fields, but should have been leaving them `None` to spend his `P2WPKH` Input. See the commented code below:

Image

I believe I recall @konstantinullrich having a similar problem in implementation.

I wonder if AddressType-specific constructors (e.g. `InputPair::new_p2tr`, `InputPair::new_p2wpkh` etc.) would have prevented this implementation error. It was tricky to hunt down what the problem even was since it only failed script verification.

To be explicit, I think their [flutter (dart) implementation](https://github.com/SatoshiPortal/bullbitcoin-mobile/blob/420f13819c75f1c149878ee0cf9217c96989839a/lib/_pkg/payjoin/manager.dart#L740-L757):

```dart
Future<InputPair> _inputPairFromUtxo(bdk.LocalUtxo utxo, bool isTestnet) async {
  final psbtin = PsbtInput(
    witnessUtxo: TxOut(
      value: utxo.txout.value,
      scriptPubkey: utxo.txout.scriptPubkey.bytes,
    ),
  );
  final txin = TxIn(
    previousOutput:
        OutPoint(txid: utxo.outpoint.txid, vout: utxo.outpoint.vout),
    scriptSig: await Script.newInstance(rawOutputScript: []),
    sequence: 0xFFFFFFFF,
    witness: [],
  );
  return InputPair.newInstance(txin, psbtin);
}
```

May prevent misuse if an imagined `InputPair` constructor were instead applied as

```dart
InputPair _inputPairFromUtxo(bdk.LocalUtxo utxo, bool isTestnet) {
  return InputPair.newP2wpkh(witnessUtxo: utxo.txout, previousOutput: utxo.outpoint);
}
```

and provided the other fields as defaults (perhaps with optional sequence), or further specified the necessary parameters to create a p2wpkh InputPair.

### PR DESCRIPTION
**Pull request by Johnosezele**
Add unit tests for the new_p2wpkh and new_p2tr constructors within the `receive` module (payjoin/src/receive/mod.rs). These tests verify that the constructors correctly initialize the InputPair struct, ensuring both the `TxIn` and `psbt::Input` components match expected values.

This improves code quality by confirming constructor behavior and guarding against regressions.

Also addresses compilation errors found during testing:
- Corrected call to new_p2wpkh to include the `sequence` argument.
- Removed assertions checking non-existent fields on `psbt::Input` (e.g., `script_sig`, `tap_control_block`).
- Renamed `tap_leaf_scripts` to the correct field `tap_scripts` in assertions.

### CODE REVIEW

File: payjoin/src/receive/mod.rs [hunk 1]
```diff
@@ -57,6 +57,80 @@ impl InputPair {
...
+            tap_merkle_root: None,
+            proprietary: Default::default(),
+            unknown: Default::default(),
+            partial_sigs: Default::default(),
+            sighash_type: None,
+        };
+        Self { txin, psbtin }
+    }
+
+    pub fn new_p2tr(
```
Comment (benalleng) [inline payjoin/src/receive/mod.rs:97]: Is there a reason to not deduplicate these constructors as just `new fn input_pair`?

Comment (benalleng) [reply payjoin/src/receive/mod.rs:97]: Ok I see this referenced in the issue, I guess it would just be nice to have relevant params to give them some reason to be different

File: payjoin/src/receive/mod.rs [hunk 2]
```diff
@@ -57,6 +57,80 @@ impl InputPair {
...
+            tap_internal_key: None,
+            tap_merkle_root: None,
+            proprietary: Default::default(),
+            unknown: Default::default(),
+            partial_sigs: Default::default(),
+            sighash_type: None,
+        };
+        Self { txin, psbtin }
+    }
+
```
Comment (spacebear21) [inline payjoin/src/receive/mod.rs:133]: We should offer constructors for at least every input type supported by [`expected_input_weight`](https://github.com/spacebear21/rust-payjoin/blob/master/payjoin/src/psbt/mod.rs#L184), so p2pkh and nested p2wpkh-in-p2sh constructors must be available and tested as well.

File: payjoin/src/receive/mod.rs [hunk 3]
```diff
@@ -87,3 +161,94 @@ pub(crate) fn parse_payload(
...
+        assert_eq!(input_pair.psbtin.tap_internal_key, None); // P2TR specific check
+        assert_eq!(input_pair.psbtin.tap_merkle_root, None); // P2TR specific check
+        assert!(input_pair.psbtin.proprietary.is_empty());
+        assert!(input_pair.psbtin.unknown.is_empty());
+
+        // Check txin field
+        assert_eq!(input_pair.txin.previous_output, dummy_outpoint);
+        assert_eq!(input_pair.txin.script_sig, ScriptBuf::new()); // Should be empty for P2TR
+        assert_eq!(input_pair.txin.sequence, dummy_sequence); // Check sequence default
+        assert!(input_pair.txin.witness.is_empty()); // Witness should be empty initially
```
Comment (spacebear21) [inline payjoin/src/receive/mod.rs:252]: I don't think checking every default field is useful, I'd rather see us call relevant functions that use `InputPair` such as `expected_input_weight` and ensure they succeed as expected.

File: payjoin/src/receive/mod.rs [hunk 4]
```diff
@@ -57,6 +57,80 @@ impl InputPair {
...
+            tap_scripts: Default::default(),
+            tap_key_origins: Default::default(),
+            tap_internal_key: None,
+            tap_merkle_root: None,
+            proprietary: Default::default(),
+            unknown: Default::default(),
+            partial_sigs: Default::default(),
+            sighash_type: None,
+        };
+        Self { txin, psbtin }
```
Comment (spacebear21) [inline payjoin/src/receive/mod.rs:94]: It's critical that these InputPair's be validated, which is what the `new` constructor does.  Perhaps `new` should be a private function that gets called in turn by these helper constructors.

**Review by spacebear21 — CHANGES_REQUESTED** [review]

File: payjoin/src/receive/mod.rs (hunk 1 above)
Comment (Johnosezele) [reply payjoin/src/receive/mod.rs:97]: Thanks for the suggestion! Yes, while the constructors currently share similar parameters, keeping them separate allows us to:
- Make the API more explicit about which type of input is being created
- Add Taproot-specific parameters to new_p2tr in the future (like tap internal keys or merkle roots) without breaking changes
- Keep P2TR functionality properly gated behind the v2 feature flag

File: payjoin/src/receive/mod.rs [hunk 5]
```diff
@@ -57,6 +57,80 @@ impl InputPair {
...
+            tap_script_sigs: Default::default(),
+            tap_scripts: Default::default(),
+            tap_key_origins: Default::default(),
+            tap_internal_key: None,
+            tap_merkle_root: None,
+            proprietary: Default::default(),
+            unknown: Default::default(),
+            partial_sigs: Default::default(),
+            sighash_type: None,
+        };
```
[the remark this answers is not in the document]
Comment (Johnosezele) [reply payjoin/src/receive/mod.rs:93]: Thanks for the suggestion! You're right - using ..Default::default() is much cleaner and achieves the same result since we're setting everything except witness_utxo to its default value. I've updated both constructor functions to use this pattern. Makes it even more maintainable and less prone to errors if new fields are added to psbt::Input in the future.

File: payjoin/src/receive/mod.rs [hunk 6]
```diff
@@ -57,6 +57,80 @@ impl InputPair {
...
+        witness_utxo: TxOut,
+        previous_output: OutPoint,
+        sequence: Option<Sequence>,
+    ) -> Self {
+        let txin = TxIn {
+            previous_output,
+            script_sig: ScriptBuf::new(),
+            sequence: sequence.unwrap_or_default(),
+            witness: Witness::default(),
+        };
```
[the remark this answers is not in the document]
Comment (Johnosezele) [reply payjoin/src/receive/mod.rs:70]: Fixed too 🫡

[the remark this answers is not in the document]
Comment (Johnosezele) [reply payjoin/src/receive/mod.rs:70]: I've implemented them locally and verified all tests pass.

File: payjoin/src/receive/mod.rs (hunk 2 above)
Comment (Johnosezele) [reply payjoin/src/receive/mod.rs:133]: Thanks for the suggestion! I've added constructors for:
-  new_p2pkh() for legacy inputs
-  new_p2sh_p2wpkh() for nested SegWit inputs
I noticed that expected_input_weight() explicitly returns NotSupported for P2WSH inputs. Should we:
-  Keep P2WSH unsupported (current behavior)
-  Add P2WSH support to both expected_input_weight() and create a new new_p2wsh() constructor
What do you think?

File: payjoin/src/receive/mod.rs (hunk 3 above)
Comment (Johnosezele) [reply payjoin/src/receive/mod.rs:252]: Yeah, true its not useful, I should test behavior instead of implementation. I have added test for [additional_input_weight](https://github.com/spacebear21/rust-payjoin/blob/55a4b7ee15fd31fe88392147adf75fca94ffde50/payjoin/src/receive/v1/mod.rs#L646). I've added test that verifies the relative weights of different input types (P2WPKH vs P2PKH), which is crucial for the fee calculation logic. This test ensures that the weight calculations that feed into fee estimation are working correctly.

Comment (Johnosezele) [reply payjoin/src/receive/mod.rs:252]: This test verifies that P2WPKH inputs have lower weight than P2PKH inputs, which is one of the key features of witness transactions.

Comment (Johnosezele) [reply payjoin/src/receive/mod.rs:252]: Hi @spacebear21 I've implemented comprehensive tests for the key privacy and fee-related behaviors in the payjoin protocol. Here's a breakdown of the test coverage:

### Privacy-Based Input Selection Test
The [test_privacy_based_input_selection](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:399:4-482:5) test verifies the core privacy feature of payjoin - avoiding the Unnecessary Input Heuristic (UIH). It:
- Creates two candidate inputs with different amounts (5,000 sats and 50,000 sats)
- Verifies that `try_preserving_privacy` selects an input that maintains privacy
- Tests that the selection logic properly considers input amounts to avoid revealing patterns
- Ensures the receiver's change output is properly handled to maintain privacy

This test is crucial for validating that our implementation effectively prevents blockchain analysis heuristics.

### Fee Rate Enforcement Test
The [test_fee_rate_based_selection](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:484:4-506:5) test focuses on security against probing attacks through fee rate enforcement:
- Sets a minimum fee rate of 2 sat/vB in the proposal
- Verifies that `check_broadcast_suitability` rejects proposals below the minimum fee rate
- Tests the fee rate validation logic in the protocol
- Ensures proper handling of maximum effective fee rates

This helps prevent potential probing attacks by enforcing appropriate fee rates.

### Fee Handling Test
The [test_minimum_fee_requirements](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:508:4-584:5) test validates the complete fee handling flow:
- Tests fee application when contributing inputs
- Verifies fees can be successfully applied within specified min/max ranges
- Uses `finalize_proposal` (the public API) to test the complete proposal flow
- Ensures the receiver can properly cover their own input costs while maintaining protocol security

I've made sure to test both the happy path and edge cases for each of these behaviors. Let me know if you'd like me to add any additional test cases or if you have any suggestions for improving the test coverage for my next pr.

File: payjoin/src/receive/mod.rs (hunk 4 above)
Comment (Johnosezele) [reply payjoin/src/receive/mod.rs:94]: Good point about the [new](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:77:4-86:5) constructor! I agree that making it private would better encapsulate the validation logic. Here's what I propose:

1. Make [new](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:77:4-86:5) private (remove `pub`) since it's an internal implementation detail
2. Keep the specialized constructors ([new_p2wpkh](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:88:4-103:5), [new_p2pkh](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:105:4-120:5), etc.) public as they provide a safer, more guided interface
3. These specialized constructors will internally call the private [new](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:77:4-86:5) function after setting up the correct PSBT inputs

This way:
- We maintain strict validation through the private [new](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:77:4-86:5) function
- Users are guided to use the appropriate constructor for their address type
- We prevent potential misuse by requiring proper UTXO and script setup through the specialized constructors

Would you like me to make these changes in a follow-up PR?

Comment (Johnosezele) [reply payjoin/src/receive/mod.rs:94]: You make a great point about validation! I initially made [new](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin-cli/src/app/wallet.rs:23:4-36:5) private but realized it broke the CLI's usage. After some thought, I have a better approach:

1. Make [new](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin-cli/src/app/wallet.rs:23:4-36:5) private (remove `pub`) since it's an internal implementation detail
2. Keep the specialized constructors ([new_p2wpkh](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:88:4-103:5), [new_p2pkh](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin/src/receive/mod.rs:105:4-120:5), etc.) public as they provide a safer, more guided interface
3. These specialized constructors will internally call the private [new](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin-cli/src/app/wallet.rs:23:4-36:5) function after setting up the correct PSBT inputs

Benefits of this approach:
- We maintain strict validation through the private [new](cci:1://file:///Users/mac/Dev/rust-payjoin-1/payjoin-cli/src/app/wallet.rs:23:4-36:5) function
- Users are guided to use the appropriate constructor for their address type
- We prevent potential misuse by requiring proper UTXO and script setup through the specialized constructors

I'll make these changes in a follow-up PR.

**Comment by Johnosezele** [comment]
PR Updated:
- Removed `test_new_p2wpkh` as suggested
- Cleaned up test comments for better readability
- All tests are passing
- Added comprehensive test coverage:
  - `test_privacy_based_input_selection`: Tests UIH avoidance
  - `test_fee_rate_based_selection`: Validates fee rate enforcement
  - `test_minimum_fee_requirements`: Tests end-to-end fee handling

Ready for another review! 👍

File: payjoin/src/receive/mod.rs (hunk 1 above)
Comment (benalleng) [reply payjoin/src/receive/mod.rs:188]: Ok, Its much clearer now the differences between the different descriptors when extending the defaults out.
